### DISCUSSION TITLE
UserRepresentation & Roles

### DISCUSSION THREAD

**Discussion by legitimoth** [keycloak/keycloak#10168] [Q&A]
Hey guys!

I'm using [keycloak-nodejs-admin-client](https://github.com/keycloak/keycloak-nodejs-admin-client) to communicate with keycloak API.

I'm trying to use `kcAdminClient.users.find()` to get all users and it works fine but it's not bringing any role. 

I saw the UserRepresentation's structure and there are two properties : `clientRoles?: Record<string, any>` and  `realmRoles?: string[]`

Why it's not being filled? Did I have to setup something into my administration console? 

Thx for the attention guys!

**Comment by Crafoord** [comment]
I have the same issue 2.5 years later, help!

**Comment by danielFesenmeyer** [comment]
Afaik, the properties are used when writing users only.

For retrieving user-assigned roles, you can use a subresource:  https://github.com/keycloak/keycloak/blob/db7694e7be6c600ef5ca98a29e7f65f0cc3b5d2d/js/libs/keycloak-admin-client/src/resources/users.ts#L106

**Comment by danielFesenmeyer** [comment]
Also see related discussion: https://github.com/keycloak/keycloak/discussions/30782

**Comment by Crafoord** [comment]
Thanks for the input @danielFesenmeyer, in my case I need to retrieve a lot of users with roles so it seems making a lot of requests is the only way (get 100 users = 1 request, then get roles for 100 users = 100 requests). I guess this is the way for now!
